modules/tftp/server.rb

```diff
@@ -150,6 +150,36 @@ def pxeconfig_file(mac)
     end
   end
 
+  def self.fetch_boot_image(image_dst, url, files)
+    # Verify image_dst does not contain ".." (switch folder)
+    raise "TFTP image file extraction error: #{file_path}" unless image_dst.split("..").size <= 1
+
+    base_path = Pathname.new("/var/www/html/pub/installation_media")
+    image_path = Pathname.new(File.expand_path(image_dst, base_path)).cleanpath
+    extr_image_dir = Pathname.new(image_path.to_s.delete_suffix(".iso"))
+
+    FileUtils.mkdir_p image_path.parent
+    download_task = choose_protocol_and_fetch(url, image_path)
+    # Wait for concurrent processes
+    raise "Cannot download boot image at the moment - is another process downloading it already?" if download_task.is_a?(FalseClass)
+    # wait for download completion
+    download_task.join
+
+    files.each do |file|
+      raise "File to extract from image contains up-directory: #{file_path}" unless file.split("..").size <= 1
```

Comment:
Isn't this the same as
```suggestion
      raise "File to extract from image contains up-directory: #{file_path}" unless file.include?("..")
```

This can actually be a false positive. Take `/path/to/some...file.iso'` for example. This is a valid path without exploits.

A better way would be to see if the absolute path of `extr_file_path` is a subdirectory of `extr_image_dir`. I think this works:
```
if extr_file_path.expand_path.relative_path_from(extr_image_dir).to_s.start_with('..')
  raise "File to extract from image contains up-directory: #{file_path}"
end
```

There may be a shorter or more efficient way.

Reply:
Done, thanks for the hint!